src/lifo.zig

```diff
@@ -0,0 +1,216 @@
+        }
+
+        /// Returns the first element of the stack, and removes it.
+        pub fn pop(self: *LIFO) ?*T {
+            assert((self.count == 0) == (self.head == null));
+
+            const node = self.head orelse return null;
+            self.head = node.next;
+            node.next = null;
+            self.count -= 1;
+            return node;
+        }
+
+        /// Returns the first element of the stack, but does not remove it.
+        pub fn peek(self: LIFO) ?*T {
+            return self.head;
+        }
+
+        /// Checks if the LIFO is empty.
+        pub fn empty(self: LIFO) bool {
+            assert((self.count == 0) == (self.head == null));
+            return self.head == null;
+        }
+
+        /// Returns whether the linked list contains the given *exact element* (pointer comparison).
+        fn contains(self: *const LIFO, needle: *const T) bool {
+            var iterator = self.head;
+            var count: u64 = 1;
+            while (iterator) |node| : (iterator = node.next) {
+                if (node == needle) return true;
+                count += 1;
+                assert(count <= self.count_max);
+            }
+            return false;
+        }
+
+        /// Resets the state.
+        pub fn reset(self: *LIFO) void {
+            self.* = .{
+                .name = self.name,
+                .count_max = self.count_max,
+                .verify_push = self.verify_push,
+            };
+        }
+    };
+}
+
+test "LIFO: fuzz" {
+    // Fuzzy test to compare behavioud of LIFO against ArrayList (reference model).
+    comptime assert(constants.verify);
+
+    const fuzz = @import("testing/fuzz.zig");
+    const allocator = std.testing.allocator;
+
+    var prng = std.Random.DefaultPrng.init(0);
+    const random = prng.random();
+
+    const Node = struct {
+        id: u32,
+        next: ?*@This() = null,
+    };
+    const LIFO = LIFOType(Node);
+
+    const node_count_max = 1024;
+    const events_max = 1 << 8;
+
+    const Event = enum { push, pop };
+    const event_distribution = fuzz.DistributionType(Event){
+        .push = 2,
+        .pop = 1,
+    };
+
+    // Allocate a pool of nodes.
+    var nodes = try allocator.alloc(Node, node_count_max);
+    defer allocator.free(nodes);
+    for (nodes, 0..) |*node, i| {
+        node.* = Node{ .id = @intCast(i), .next = null };
+    }
+
+    // A bit set that tracks which nodes are available.
+    var nodes_free = try std.DynamicBitSetUnmanaged.initFull(allocator, node_count_max);
+    defer nodes_free.deinit(allocator);
+
+    var stack = LIFO.init(node_count_max, "fuzz", true);
+
+    // Reference model: a dynamic array of node IDs in LIFO order (last is the top).
+    var model = try std.ArrayList(u32).initCapacity(allocator, node_count_max);
+    defer model.deinit();
+
+    // Run a sequence of randomized events.
+    for (0..events_max) |_| {
+        std.debug.assert(model.items.len <= node_count_max);
```

Comment:
```suggestion
        assert(model.items.len <= node_count_max);
```

We in general are all other the place regarding qualified imports, but `assert` is one instance where we _always_ use it unqualified!